[14.0][MIG] report_py3o: Migration to 14.0

elmeriniemela opened the pull request.

lmignon committed:
Move to report_py3o after import from HG

lmignon committed:
Add setup

lmignon committed:
[FIX] View inheritance may not use attribute 'string' as a selector.

lmignon committed:
[IMP] New API + Short Hearder

lmignon committed:
Remove unicode string

lmignon committed:
[IMP] Add the possiblity to get the template from an absolute path on the server

lmignon committed:
[IMP] Allow user to use local fusion

lmignon committed:
Add a simple way to extend the parser context

lmignon committed:
Remove ACL on removed model

lmignon committed:
Improve filetype selection name

lmignon committed:
Fix native flag

faide committed:
fixed missing comma in model definition

lmignon committed:
Add simple tests and fix call to fusion server

lmignon committed:
Fix skip fusion on server when the fusion is local

lmignon committed:
Improve tests

lmignon committed:
Replace openerp element by odoo

alexis-via committed:
Small usability improvements

Replace README.md by README.rst (not finished)

alexis-via committed:
Take into accounts most remarks of @lasley

Remove <data> in views
Protect import of py3o libs
Remove dep on base module
Other small changes

lmignon committed:
Replace plain SQL by orm

lmignon committed:
rename module ir_report to ir_actions_report_xml

lmignon committed:
Add more tests and fixes issues found by tests

lmignon committed:
[FIX] Check constrains only if report_type == 'py3o'

alexis-via committed:
report_py3o: full installation procedure in README

alexis-via committed:
Port report_py3o to Odoo v10

alexis-via committed:
Update version numbers

Rename __openerp__.py to __manifest__.py

alexis-via committed:
PEP8 fix

alexis-via committed:
Update README

lmignon committed:
[IMP] Escape 'False' when rendering value

lmignon committed:
[IMP] Replace old style parser by TransientModel
The goal is to improve the modularity by making the parser a true inheritable odoo model and share part of the code with the 'report' model

Conflicts:
	report_py3o/models/ir_actions_report_xml.py
	report_py3o/models/py3o_report.py
	report_py3o/tests/test_report_py3o.py

alexis-via committed:
FIX double dot in filename

alexis-via committed:
Add access to report_xml object in report

alexis-via committed:
FIX Adapt test suite to my recent commit to avoid double dot on filename extension

Commit:
[FIX] imports

lmignon committed:
[IMP] Minimizes memory consumption

Conflicts:
	report_py3o/models/py3o_report.py

Commit:
* travis.yml
* flake8

alexis-via committed:
Update and improve README instructions to setup py3o

lmignon committed:
[IMP] Allow to override/extend the way we get the fallback template

lmignon committed:
[IMP] report_py3o: Take into account print_report_name
If a Printed Report Name is set on the action report, use it as
downloaded filename
refs #133

lmignon committed:
[FIX] report_py3o: fix exception when report must be saved as attachement
The mehtod  must be called with a list of ids not with a list of browse records

oca-transbot committed:
OCA Transbot updated translations from Transifex

alexis-via committed:
[FIX] delete PDF invoice attachment on invoice back to draft (native feature now working with py3o)

lmignon committed:
[IMP] report_py3o: update authors and contributors

lmignon committed:
[IMP] report_py3o: prevent injections when retrieving the template from path

oca-transbot committed:
OCA Transbot updated translations from Transifex

oca-transbot committed:
OCA Transbot updated translations from Transifex

sbidoul committed:
report_py3o: bump version following last merge

Commit:
Fix py3o freeze when using server fusion

The template file has been partially read for the parsing of the
expressions, so use the original template data instead.  When the fusion
server receives an incomplete template, we don't receive any answer
back.

Commit:
Pass option for escaping False values to server

Needs https://bitbucket.org/faide/py3o.fusion/pull-requests/1

alexis-via committed:
Add option py3o_multi_in_one for Py3o reports

oca-transbot committed:
OCA Transbot updated translations from Transifex

lmignon committed:
[FIX] report_py3o:  must return the path to the report
refs #119

oca-transbot committed:
OCA Transbot updated translations from Transifex

omar7r committed:
[IMP] report_py3o: Allows to interpret \n and \t in texts, cleans html tags too

omar7r committed:
[FIX] Deletes an unnecessary function call

omar7r committed:
[FIX] report_py3o: Removes api.one

omar7r committed:
[FIX] report_py3o: Import new dependencies in try...except

omar7r committed:
[UPD] report_py3o: Bump version

oca-transbot committed:
OCA Transbot updated translations from Transifex

hbrunn committed:
[RFR] split off the fusion server to its own module

use libreoffice for conversions in the base version. Fixes #179

jesusVMayor committed:
[FIX]report_py3o: Escape correctly html characters.

alexis-via committed:
Improve README.rst for report_py3o* modules

hbrunn committed:
[ADD] migration script to install report_py3o_fusion_server on upgrades

oca-transbot committed:
OCA Transbot updated translations from Transifex

hbrunn committed:
[FIX] fallback to libreoffice conversion when no server is configured

fixes #200

pedrobaeza committed:
[FIX] report_py3o: Bump version number

hbrunn committed:
[IMP] check for template data, not for the id

alexis-via committed:
report_py3o_fusion_server: Add support for PDF Export options of libreoffice

SimoRubi committed:
[FIX] If this is not a py3o report, do nothing at all. In some cases the UI remains blocked otherwise

sbidoul committed:
report_py3o: mention formatLang in README

oca-travis committed:
[UPD] Update report_py3o.pot

kryskool committed:
Translated using Weblate (Français)

Currently translated at 100,0% (38 of 38 strings)

Translation: reporting-engine-10.0/reporting-engine-10.0-report_py3o
Translate-URL: https://translation.odoo-community.org/projects/reporting-engine-10-0/reporting-engine-10-0-report_py3o/fr/

lmignon committed:
[MIG] report_py3o, report_py3o_fusion_server: Migration to 12.0

lmignon committed:
[IMP] report_py2o: cleanup js

Commit:
[12.0][MIG] improvement py3o_report_extender

sbejaoui committed:
Update i18n

lmignon committed:
[FIX] report_py3o: Use the right naming convention for the render method
To determine the method to use to render a recport according to its type, the generic method 'render' defined into ir.action.report check if a method name  is defined on the model. https://github.com/odoo/odoo/blob/12.0/odoo/addons/base/models/ir_actions_report.py#L734
Therefore, we must provide this method to be compliant with what's expected by Odoo.
This change will also allows the usage of py3o template with mail_template once https://github.com/odoo/odoo/pull/30013 will be merged

lmignon committed:
[IMP] report_py3o: remove obsolete methods

lmignon committed:
[IMP] report_py3o: Document new methods available into the parser context

lmignon committed:
[FIX] report_py3o: Add missing method _merge_pdf.
This method was previously provided by Odoo and is used to merge all the reports generated if the generation is called for more than one record

lmignon committed:
[IMP][FIX] py3o_report, py3o_report_fusion_server: Compute the availability of py3o report
Before this change it was not possible to install modules declaring py3o report into a non native format without specifying a Fusion server once the module py3o_report_fusion_server was installed. With theses changes, we now take care of the availability of the libreoffice runtime to display/log a warning message when the report is in a non native runtime.

lmignon committed:
[FIX] report_py3o: Add missing 'user' and 'lang' into the parser context

lmignon committed:
[IMP] py3o_report: Allow to specify the lang when calling o_format_lang
lang_code is already supported by o_format_date

lmignon committed:
[FIX] report_py3o: Access to ir.config_parameter with sudo

oca-travis committed:
[UPD] Update report_py3o.pot

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: reporting-engine-12.0/reporting-engine-12.0-report_py3o
Translate-URL: https://translation.odoo-community.org/projects/reporting-engine-12-0/reporting-engine-12-0-report_py3o/

oca-travis committed:
[UPD] Update report_py3o.pot

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: reporting-engine-12.0/reporting-engine-12.0-report_py3o
Translate-URL: https://translation.odoo-community.org/projects/reporting-engine-12-0/reporting-engine-12-0-report_py3o/

alexis-via committed:
Update README for report_py3o and report_py3o_fusion_server (#306)

* Update README for report_py3o and report_py3o_fusion_server

Convert README to the new readme subdirectory

sbidoul committed:
[FIX] report_py3o readme structure

OCA-git-bot committed:
[UPD] README.rst

dupski committed:
Make sure value is always set for computed fields
(without these a CacheMiss error is thrown by odoo)

lmignon committed:
[IMP] report_py3o, report_py3o_fusion_server: black, isort

lmignon committed:
[MIG] report_py3o, report_py3o_fusion_server: Migration to 13.0

sbejaoui committed:
[12.0][FIX] - report_py3o: run libreoffice in an isolated user installation

Bug when more than one conversion print is launched within the same libreoffice instance.

The standard behavior of libreoffice when a user open it while another instance is
running is to show a new window and throw an error if a new instance is forced within
the same user installation [see](https://bugs.documentfoundation.org/show_bug.cgi?id=37531).

This implies a bug in report_py3o module when we call libreoffice at the same time for
different documents.

To reproduce this bug:

**Case 1:**
1.  Simultaneously print two documents.

**Case 2:**
1.  Run print jobs using job_queue module
2.  Manually print another document

**Case 3:**
2.  Open libreoffice
3.  Print a py3o report

This PR creates a temporary user installation for each libreoffice conversion to bypass this limitation.

lmignon committed:
[FIX] report_py3o: prettier xml after merge with master

sbidoul committed:
[FIX] make prettier happy

oca-travis committed:
[UPD] Update report_py3o.pot

OCA-git-bot committed:
[UPD] README.rst

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: reporting-engine-13.0/reporting-engine-13.0-report_py3o
Translate-URL: https://translation.odoo-community.org/projects/reporting-engine-13-0/reporting-engine-13-0-report_py3o/

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: reporting-engine-13.0/reporting-engine-13.0-report_py3o
Translate-URL: https://translation.odoo-community.org/projects/reporting-engine-13-0/reporting-engine-13-0-report_py3o/

elmeriniemela committed:
[14.0][MIG] report_py3o: Migration to 14.0

rvalyi commented:
Thank you for the work @elmeriniemela  , but please enforce the standard migration process here https://github.com/OCA/maintainer-tools/wiki/Migration-to-version-14.0 (that is mostly preserve the commit history). Please redo the pull request in the proper way, then we will be happy to review.

elmeriniemela commented:
> Thank you for the work @elmeriniemela , but please enforce the standard migration process here https://github.com/OCA/maintainer-tools/wiki/Migration-to-version-14.0 (that is mostly preserve the commit history). Please redo the pull request in the proper way, then we will be happy to review.

@rvalyi I added the git history, thanks for the tip!

mourad-ehm commented on report_py3o/models/ir_actions_report.py:
```diff
@@ -0,0 +1,200 @@
+        report = self.get_from_report_name(self.report_name, self.report_type)
+        if report.print_report_name and not len(res_ids) > 1:
+            obj = self.env[self.model].browse(res_ids)
+            return safe_eval(report.print_report_name, {"object": obj, "time": time})
```
```suggestion
            return safe_eval(report.print_report_name, {"object": obj,})
```
In v 14.0 time module is already in safe_eval https://github.com/odoo/odoo/blob/14.0/odoo/tools/safe_eval.py#L38

hparfr replied:
retreturn ?

mourad-ehm replied:
fixed. thanks

elmeriniemela committed:
Update report_py3o/models/ir_actions_report.py

In v 14.0 time module is already in safe_eval https://github.com/odoo/odoo/blob/14.0/odoo/tools/safe_eval.py#L38

Co-authored-by: mourad-ehm <mourad.elhadj.mimoune@akretion.com>

hparfr reviewed (approved):
LGTM

alexis-via commented:
It doesn't work when print_report_name is set. I fixed the bug in this commit: https://github.com/akretion/reporting-engine/commit/f10b5647ca7a37816fb8b3bc689e7bbcda867450
@elmeriniemela I wanted to make a PR to your branch, but I can't select it (maybe some special settings on your side ??). So please take the code in my commit and push it here.

elmeriniemela committed:
[MIG] Use safe_eval wrapped time module. Fix deprecation warnings.

Co-authored-by: Alexis de Lattre <alexis.delattre@akretion.com>

elmeriniemela commented:
Thanks @alexis-via ! I added your fix and made you a co-author. Also fixed some deprecation warnings.

Asemtalat commented:
> _No description provided._

Hello 
i wand v14 from that medule 
**report_py3o** 

Asemtalat commented:
if anyone have thats medule (https://apps.openerp.com/apps/modules/13.0/report_py3o)    on v 14 plz send it to my mail (asemasem728@gmail.com)

StefanRijnhart commented:
@elmeriniemela thank you for the port! Travis fails on the following error
```
FAIL: TestReportPy3o.test_py3o_report_availability

Traceback (most recent call last):

  File "/home/travis/build/OCA/reporting-engine/setup/report_py3o/odoo/addons/report_py3o/tests/test_report_py3o.py", line 228, in test_py3o_report_availability

    self.assertTrue(self.report.lo_bin_path)

AssertionError: False is not true
```
Please have a look and see if you can figure out what's wrong.

etobella commented:
libreoffice is missing on the travis configuration. 

https://github.com/OCA/reporting-engine/blob/13.0/.travis.yml#L15

Can you add the needed libraries on .travis.yml?

elmeriniemela committed:
[MIG] Fix libreoffice missing on the travis configuration

etobella commented:
Review precommit please.

Also, there is a warning related to genshi, but I don't know where it is generated
```
2021-02-10 14:46:43,478 160 WARNING openerp_test py.warnings: /.repo_requirements/virtualenv/python3.6/lib/python3.6/site-packages/genshi/template/interpolation.py:36: DeprecationWarning: Flags not at the start of the expression '[uU]?[rR]?("""|\\\'\\\'\\' (truncated)
PseudoToken
```

njeudy reviewed (approved):
Ok test, and using it in prod :)

rbr-sacatec commented:
This is not working on windows. The following error message is shown when trying to print a py3o report:

> The application cannot be started. 
> The configuration file "C:\Program Files\LibreOffice\program\bootstrap.ini" is corrupt.

I've found this open issue on libreoffice-convert repository:
[https://github.com/elwerene/libreoffice-convert/issues/32](https://github.com/elwerene/libreoffice-convert/issues/32)

The error seems to be caused by the newly added `-env:UserInstallation=file` command line option, which currently shows an error message on windows.

I suggest this change on py3o_report.py line 272 to fix the error:
```py
if user_installation and os.name != 'nt':
    cmd.append("-env:UserInstallation=file:%s" % user_installation)
```

elmeriniemela commented:
> Review precommit please.
> 
> Also, there is a warning related to genshi, but I don't know where it is generated
> 
> ```
> 2021-02-10 14:46:43,478 160 WARNING openerp_test py.warnings: /.repo_requirements/virtualenv/python3.6/lib/python3.6/site-packages/genshi/template/interpolation.py:36: DeprecationWarning: Flags not at the start of the expression '[uU]?[rR]?("""|\\\'\\\'\\' (truncated)
> PseudoToken
> ```

https://github.com/edgewall/genshi/issues/15

alexis-via commented:
To be able to include properly formatted datetime field in a report displaying date and time, I propose this patch : https://people.via.ecp.fr/~alexis/py3o-format-datetime.diff
@elmeriniemela Could you include it? I would love to make a PR to your branch, but I can't select it for some reasons.

alexis-via commented:
@elmeriniemela Could you make your branch public so that we can push a PR to your branch, to help you on this PR? Or do you prefer that we superseed this PR?
I'd like to push this PR to get it merged because I have 4 other OCA PR that are waiting for this one to be merged.

elmeriniemela commented:
@alexis-via I'm not an expert on github settings and I could not find with ~15mins of searching whats the issue with my branch. If you know what I need to change I would be happy to make the branch public. If you don't know the root cause of this branch issue either, I could make you a collaborator of the fork?

elmeriniemela committed:
[IMP] Add ability to include properly formatted datetime field in a report displaying date and time

elmeriniemela commented:
@alexis-via I applied your patch. (just plain git apply and commit, I didnt test it myself)

Commit:
Merge branch '14.0' into 14.0

nilshamerlinck commented on report_py3o/models/py3o_report.py:
```diff
@@ -0,0 +1,370 @@
+logger = logging.getLogger(__name__)
+
+try:
+    from py3o.template import Template
```
```suggestion
    # workaround for https://github.com/edgewall/genshi/issues/15
    # that makes runbot build red because of the DeprecationWarning
    with warnings.catch_warnings():
        warnings.filterwarnings("ignore", category=DeprecationWarning)
        from py3o.template import Template
```

nilshamerlinck commented on report_py3o/models/py3o_report.py:
```diff
@@ -0,0 +1,370 @@
+import os
+import subprocess
+import sys
+import tempfile
```
```suggestion
import tempfile
import warnings
```

nilshamerlinck commented on report_py3o/models/py3o_report.py:
```diff
@@ -0,0 +1,370 @@
+        for temporary_file in temporary_files:
+            try:
+                os.unlink(temporary_file)
+            except (OSError, IOError):
```
```suggestion
            except OSError:
```

nilshamerlinck commented:
Hi @elmeriniemela applying the suggested changes + running `pre-commit run -a` will make both checks green :)

elmeriniemela committed:
[IMP] Workaround for genshi DeprecationWarning that makes runbot build red

Co-authored-by: Nils Hamerlinck <nilshamerlinck@users.noreply.github.com>

elmeriniemela committed:
[IMP] Cleanup unnecessary imports

Co-authored-by: Nils Hamerlinck <nilshamerlinck@users.noreply.github.com>

elmeriniemela committed:
[IMP] Remove except of IOError as its an alias to OSError in python3

Co-authored-by: Nils Hamerlinck <nilshamerlinck@users.noreply.github.com>

lmignon commented:
@elmeriniemela Thank you for all this work. Can you check the pre-commit issues https://github.com/OCA/reporting-engine/runs/4149149084?check_suite_focus=true#step:4:106 (mainly black problem).Once fixed I'll merge your PR

lmignon reviewed (changes requested):
pre-commit issues to fix. Otherwise LGTM (code review only)

leemannd committed:
run pre-commit

leemannd commented:
Hello @elmeriniemela you can fix your PR by running `pre-commit run -a` or either merge the PR I did against your branch https://github.com/elmeriniemela/reporting-engine/pull/1

elmeriniemela committed:
Merge pull request #1 from leemannd/14.0-mig-report_py3o

run pre-commit

leemannd commented:
@OCA/reporting-engine-maintainers Ready to be merged :)

leemannd commented:
Travis is failing because it is trying to fetch a closed PR with the same code. I'm not sure why it does so.

nilshamerlinck commented:
Looks like a bug on travis side indeed. Given that the travis build was green before and pre-commit cleaning has now been done, I think this can be merged as is @lmignon @alexis-via :)

lmignon commented:
@nilshamerlinck Thank you for all your work. I've restarted travis jobs. We need to have travis :green_circle: to be able to merge.

lmignon commented:
/ocabot merge nobump

lmignon commented:
Strange... For travis, this pr seems to be linked to #551
